File: Python/Numbers/Smallest number with sum of digits as N and divisible by 10^N.py
#Problem link:https://practice.geeksforgeeks.org/problems/smallest-number-with-sum-of-digits-as-n-and-divisible-by-10n4032/1

#Time Complexity: O(N)
#Space Complexity: O(1)

class Solution:
	def digitsNum(self, N):
	    
      #to store the final number
	    s=""
	    
      #for a number to be divisible
      #by 10^N, it needs at least N zeroes
      #at the end of the number, to be
      #the smallest it needs exactly N
	    for i in range(N):
	        s = s + "0"
        
        while N > 0:
            temp = 0
          
            #We insert 9 at the start of number
            #until the sum doesn't exceed N
            if N > 9:
                s = "9" + s
                temp = 9
            #if any remainder left, add it
            #at the beginning of the number to
            #make it the smallest
            else:
                s = str(N) + s
                temp = N

            N = N - temp
            
        return s

File: Java/Numbers/Smallest number with sum of digits as N and divisible by 10^N.java
// Problem Link: https://practice.geeksforgeeks.org/problems/smallest-number-with-sum-of-digits-as-n-and-divisible-by-10n4032/1


/***
 * Approach: need to find smallest number= sumOfDigits(N) and divisble by (10^N)
 * 1. Any number that is divisible by N, should be have N zeroes.
 * 2. if any number that can be written with sumOfDigits,
 * while (N>9) N=N-9 and add "9"
 * and then, add that N. this will surely create a smallest number
 */

class Solution
{
    public String digitsNum(int N)
    {
        // Code here
        String s="";
        for (int i=0;i<N;i++){
            s=s+"0";
        }
        while(N>0){
            int temp=0;
            if(N>9){
                s="9"+s;
                temp=9;
            }else{
                s=N+s;
                temp=N;
            }
            N=N-temp;
        }
        return s;
    }
}